### PR TITLE
Move to settings v5 and rdctl API v1

### LINKED ISSUE #3746: Implement settings v5 schema [2d]
The v5 structure layout is being discussed in #3740 and the final version should be copied over into this story, once the discussion is settled.

This story is basically just a bunch of variable renames.

Please put the move from `kubernetes.containerEngine` to `containerEngine.Name` into a separate commit, because only a subset of the `containerEngine` references are affected, so this will be easier to review.

# Acceptance Criteria

| Scenario | Expected Output | 
| :---: | :---: |
|  Fresh install Rancher Desktop | check schema of created settings.json  |

### LINKED ISSUE #3747: Migrate settings.json v4 → v5 during upgrade [2d]
Add migration when Rancher Desktop is started with a v4 `settings.json` file (e.g. after an upgrade from 1.7.x).

Note: we always require a Factory Reset if the user downgrades to a previous release, so settings would be lost anyways. We don't preserve the v4 settings file.

# Acceptance criteria

* Verify that upgrading RD migrates the schema
  * Run RD 1.7.0 and set all preferences to non-default versions
  * Check current schema of settings.json(v4)
  * Upgrade to RD with the migration code
  * Verify that all values have been migrated to the new settings. 
  * Check schema of created settings.json(v5)

### LINKED ISSUE #3756: Implement new `/v1` API version for version 5 settings format [5d]
We need to bump our API version to `/v1` because the `settings` API will be changed in an incompatible way.

All `/v0` endpoints should now return an error. We don't want implement the `GET` method for `/v0/settings` (`PUT` alone is simple because we need migration code for updates anyways).

The PUT method for the `/v1/settings` endpoint will return an error when the call attempts to modify any structure that (directly, or via its parents) has a `locked` attribute that is set to `true` in the current settings.

I.e. you cannot modify `containerEngine.allowedImages.enabled` if `containerEngine.allowedImages.locked` is currently `true`.

We require that every `PUT` request of the `/v1/settings` endpoint includes the `version: 5` field in the request body. That way we could make further changes to the settings layout in the future without having to bump the API version every time.

# Acceptance criteria

* All `/v0` endpoints return an error - run `rdctl api /v0/settings` should bring an error
* The `/v1/settings` endpoint supports the new v5 schema for both `GET` and `PUT`
* The `PUT` method for `/v1/settings` returns an error unless the `version` field exists and has the value `5`.

  ```
  rdctl api -X PUT -b '{"version":5,"containerEngine":{"imageAllowList":{"enabled":true,"patterns":["a","b"]}}}' /v1/settings
  ```

# Release Notes:

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #3746
Fixes #3747
Fixes #3756

### CODE REVIEW

File: .github/workflows/macM1-e2e.yaml [hunk 1]
```diff
@@ -33,18 +33,20 @@ jobs:
...
           cp -rf $HOME/actions-runner/_work/rancher-desktop/rancher-desktop/resources/darwin/bin/ $HOME/.rd/bin/
           export PATH="$HOME/.rd/bin:$PATH"
-      - name: Suppress sudo before start up
+      - name: Disable admin-access before start up
         run: |
           mkdir -p $HOME/Library/Preferences/rancher-desktop
           touch $HOME/Library/Preferences/rancher-desktop/settings.json
           cat <<EOF > $HOME/Library/Preferences/rancher-desktop/settings.json
           {
             "version": 4,
```
Comment (mook-as) [inline .github/workflows/macM1-e2e.yaml:42]: I'd have expected this to be version 5?

File: bats/tests/helpers/vm.bash [hunk 2]
```diff
@@ -60,16 +60,18 @@ EOF
         RD_USE_IMAGE_ALLOW_LIST=true
     fi
 
-    # Make sure supressSudo is true
+    # Make sure adminAccess is false
     cat <<EOF > $RD_CONFIG_FILE
 {
   "version": 4,
```
Comment (mook-as) [inline bats/tests/helpers/vm.bash:66]: This should probably be version 5.

File: e2e/rdctl.e2e.spec.ts [hunk 3]
```diff
@@ -675,7 +678,7 @@ test.describe('Command server', () => {
           stdout, stderr, error,
         }).toEqual({
           error:  expect.any(Error),
-          stderr: expect.stringContaining(`invalid value for option --container-engine: <"${ myEngine }">; must be 'containerd', 'docker', or 'moby'`),
+          stderr: expect.stringContaining(`Error: invalid value for option --container-engine: <"${ myEngine }">; must be 'containerd', 'moby', or 'docker'`),
```
Comment (mook-as) [inline e2e/rdctl.e2e.spec.ts:713]: Oh we're not sorting it anymore? Do we need to worry about the order not being consistent?

File: pkg/rancher-desktop/backend/backend.ts [hunk 4]
```diff
@@ -74,7 +74,7 @@ export interface BackendEvents {
 /**
  * Settings that KubernetesBackend can access.
  */
-export type BackendSettings = RecursiveReadonly<Settings>;
+export type BackendSettings = Settings;
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/backend.ts:77]: Don't we want to ensure that we don't accidentally modify the passed-in settings? The previous design was ensuring that the backends had an immutable _snapshot_ of settings, and new ones only get applied via explicit calls to `start()` etc.

File: pkg/rancher-desktop/backend/lima.ts [hunk 5]
```diff
@@ -265,7 +265,10 @@ export default class LimaBackend extends events.EventEmitter implements VMBacken
 
   protected readonly CONFIG_PATH = path.join(paths.lima, '_config', `${ MACHINE_NAME }.yaml`);
 
-  protected cfg: BackendSettings | undefined;
+  /**
+   * Initialize this field to `defaultSettings` instead of `undefined` to avoid `?.` sequences everywhere we use it.
+   */
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/lima.ts:270]: But this means that we will no longer catch cases where we accidentally run with the default configuration instead of the user configuration (on startup).

File: pkg/rancher-desktop/backend/mock.ts [hunk 6]
```diff
@@ -91,7 +90,7 @@ export default class MockBackend extends events.EventEmitter implements VMExecut
 
   noModalDialogs = true;
 
-  requiresRestartReasons(config: RecursivePartial<BackendSettings>): Promise<RestartReasons> {
+  requiresRestartReasons(config: BackendSettings): Promise<RestartReasons> {
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/mock.ts:93]: I don't understand how this works: the `VMBackend` interface still calls for the argument to be a `RecursivePartial<BackendSettings>`, so the caller might pass in incomplete settings.  In fact, we _do_ do this in `background.ts` (in `proposeSettings()`) where the passed-in settings may be completely empty.

File: pkg/rancher-desktop/backend/wsl.ts [hunk 7]
```diff
@@ -132,7 +132,10 @@ export default class WSLBackend extends events.EventEmitter implements VMBackend
     return path.join(paths.resources, os.platform(), `distro-${ DISTRO_VERSION }.tar`);
   }
 
-  protected cfg: BackendSettings | undefined;
+  /**
+   * Initialize this field to `defaultSettings` instead of `undefined` to avoid `?.` sequences everywhere we use it.
+   */
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/wsl.ts:137]: See above.

File: pkg/rancher-desktop/backend/wsl.ts [hunk 8]
```diff
@@ -1432,7 +1435,7 @@ export default class WSLBackend extends events.EventEmitter implements VMBackend
   // The WSL implementation of requiresRestartReasons doesn't need to do
   // anything asynchronously; however, to match the API, we still need to return
   // a Promise.
-  requiresRestartReasons(cfg: RecursivePartial<BackendSettings>): Promise<RestartReasons> {
+  requiresRestartReasons(cfg: BackendSettings): Promise<RestartReasons> {
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/wsl.ts:1438]: See above.

File: pkg/rancher-desktop/config/__tests__/settings.spec.ts [hunk 9]
```diff
@@ -13,41 +13,45 @@ describe('updateFromCommandLine', () => {
   beforeEach(() => {
     jest.spyOn(fs, 'writeFileSync').mockImplementation(() => { });
     prefs = {
-      version:         4,
+      version:     4,
```
Comment (mook-as) [inline pkg/rancher-desktop/config/__tests__/settings.spec.ts:16]: This is clearly not version 4 anymore…

**Comment by ericpromislow** [comment]
Converting to draft -- two e2e tests are failing on windows

File: pkg/rancher-desktop/config/settings.ts [hunk 10]
```diff
@@ -21,7 +21,7 @@ const console = Logging.settings;
 // it will be picked up from the default settings object.
 // Version incrementing is for when a breaking change is introduced in the settings object.
 
-const CURRENT_SETTINGS_VERSION = 4;
+const CURRENT_SETTINGS_VERSION = 5;
```
Comment (mook-as) [inline pkg/rancher-desktop/config/settings.ts:24]: Consider doing `const CURRENT_SETTINGS_VERSION = 5 as const`, so that TypeScript will let us know when we've got the wrong version in the code. (It appears to have no ill effects on the migration code.)

File: pkg/rancher-desktop/config/settings.ts [hunk 11]
```diff
@@ -380,7 +383,7 @@ function parseSaveError(err: any) {
  * most changes will get picked up from the defaults.
  */
 const updateTable: Record<number, (settings: any) => void> = {
-  1: (settings) => {
+  1: (settings: any) => {
```
Comment (mook-as) [inline pkg/rancher-desktop/config/settings.ts:386]: Hmm, the `any` here should be redundant (because the type of `updateTable` dictates what it is).

File: pkg/rancher-desktop/config/settings.ts [hunk 12]
```diff
@@ -391,9 +394,39 @@ const updateTable: Record<number, (settings: any) => void> = {
     // The updater still wants to see an entry here (for updating ancient systems),
     // but will no longer delete obsolete files.
   },
-  3: (settings) => {
-    // Should stay true until the kim-based buildkit artifacts are removed -- see code in lima.ts:start()
-    settings.kubernetes.checkForExistingKimBuilder = true;
+  3: (_) => {
+    // With settings v5, all traces of the kim builder are gone now, so remove this line:
+    // settings.kubernetes.checkForExistingKimBuilder = true;
```
Comment (mook-as) [inline pkg/rancher-desktop/config/settings.ts:399]: Please don't leave commented code around.

Note that this means that we should either keep this as-is, or just delete the case (with a comment about why we skip 3).

File: pkg/rancher-desktop/config/settings.ts [hunk 13]
```diff
@@ -391,9 +394,39 @@ const updateTable: Record<number, (settings: any) => void> = {
...
+    delete settings.kubernetes.hostResolver;
+    delete settings.kubernetes.memoryInGB;
+    delete settings.kubernetes.numberCPUs;
+    delete settings.kubernetes.suppressSudo;
+    delete settings.kubernetes.WSLIntegrations;
+
+    delete settings.debug;
+    delete settings.pathManagementStrategy;
+    delete settings.telemetry;
+    delete settings.updater;
```
Comment (mook-as) [inline pkg/rancher-desktop/config/settings.ts:429]: We should also delete the bit about kim builder, surely?

File: e2e/rdctl.e2e.spec.ts [hunk 14]
```diff
@@ -697,6 +729,88 @@ test.describe('Command server', () => {
         });
         expect(stderr).not.toContain('Usage:');
       });
+
+      test.describe('settings v5 migration', () => {
+        /** Note issue https://github.com/rancher-sandbox/rancher-desktop/issues/3829
```
Comment (mook-as) [inline e2e/rdctl.e2e.spec.ts:734]: I believe normally `/**` is on a line by itself.

**Review by mook-as — CHANGES_REQUESTED** [review]
- The first commit is large enough that we should split out the kim builder stuff into a separate commit (or whatever else is required to make the first commit only the settings structure change + mechanical replacements elsewhere).
- Since our API is now at V1, please add a document on our API stability policy, whatever that is.

File: .github/workflows/macM1-e2e.yaml (hunk 1 above)
Comment (ericpromislow) [reply .github/workflows/macM1-e2e.yaml:42]: good catch

File: bats/tests/helpers/vm.bash (hunk 2 above)
Comment (ericpromislow) [reply bats/tests/helpers/vm.bash:51]: yup

File: e2e/rdctl.e2e.spec.ts (hunk 3 above)
Comment (ericpromislow) [reply e2e/rdctl.e2e.spec.ts]: The order comes out of the order as specified in the YAML file. So no more explicit sorting on it.

File: pkg/rancher-desktop/config/settings.ts (hunk 10 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/settings.ts:24]: ok

File: pkg/rancher-desktop/config/settings.ts (hunk 11 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/settings.ts:390]: true

File: pkg/rancher-desktop/config/settings.ts (hunk 12 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/settings.ts:403]: Leaving the explanatory comment in place, although it could have been like the change leaving version 1.

File: pkg/rancher-desktop/config/settings.ts (hunk 13 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/settings.ts:433]: yes

File: e2e/rdctl.e2e.spec.ts (hunk 14 above)
Comment (ericpromislow) [reply e2e/rdctl.e2e.spec.ts:733]: ok

File: bats/tests/helpers/vm.bash [hunk 15]
```diff
@@ -66,18 +68,19 @@ EOF
...
           3>&-
 }
 
 start_kubernetes() {
-    start_container_runtime \
-        --kubernetes-enabled \
-        --kubernetes-version "$RD_KUBERNETES_PREV_VERSION"
+    $RDCTL start \
+      --container-engine="$container_runtime" \
+      --kubernetes-enabled=false 3>&-
```
Comment (jandubois) [inline bats/tests/helpers/vm.bash:85]: Please don't make changes to bats! This will just create merge conflicts later. The current changes in this PR seem to be reverting recently merged change and look to be broken now.

I'll update BATS to use v5 in my next refactoring PR (assuming this PR gets merged first).

File: pkg/rancher-desktop/backend/backend.ts (hunk 4 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/backend/backend.ts]: I put the types back in

File: pkg/rancher-desktop/backend/lima.ts (hunk 5 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/backend/lima.ts]: Reinstated

File: pkg/rancher-desktop/backend/mock.ts (hunk 6 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/backend/mock.ts]: Reinstated

File: pkg/rancher-desktop/backend/wsl.ts (hunk 7 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/backend/wsl.ts]: Reinstated

File: pkg/rancher-desktop/backend/wsl.ts (hunk 8 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/backend/wsl.ts:1436]: Reinstated

File: pkg/rancher-desktop/config/__tests__/settings.spec.ts (hunk 9 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/__tests__/settings.spec.ts]: Yes.

File: bats/tests/helpers/vm.bash (hunk 15 above)
Comment (ericpromislow) [reply bats/tests/helpers/vm.bash:85]: ok, that happened due to a rebase conflict

**Comment by ericpromislow** [comment]
First of all, the v1 API part of this commit is actually the third of 3 PRs on the dev branch. The idea was to get the first two PRs go through a preliminary review (during which Doug did find some discrepancies -- shout out for the review process), and then put the full branch through a second review into main.

The three PRs:
* bring in the v5 schema and update all usages (like in the UI)
* migrate from v4 to v5
* move the API from /v0 to /v1 to handle the new preferences schema

Because `main` changed enough, I found it easier to rebase the third PR against `main`, so what you're seeing here is in effect a branch-review PR. I figured the v1 change didn't need a separate PR review into the dev branch, because it was mostly a string replace, with two new functions in the HTTP server, one to handle v0 requests, and the other to handle v_n_, where _n_ is > 1

Do you find it important to move the kim-builder code into a separate PR? It's a lot of work, and I don't see that it's worth it.

Summing up, there are four parts to this PR:
- new v5 schema -- most of this PR, manifested as name changes and changes of location (like `kubernetes.containerEngine` => `containerEngine.name`, etc.
- migrating to v5 -- constrained to one function in `settings.ts`
- migrating to the v1 server API -- constrained to rdctl code and a few mods in `httpCommandServer.ts`
- dropping the kim builder

I don't think this is the place to write up our policy on API stabilization, so I wrote up 3861

**Comment by jandubois** [comment]
> Do you find it important to move the kim-builder code into a separate PR?

No, it can be part of the refactoring PR that does the variable renaming.

**Comment by ericpromislow** [comment]
locked field checking will be done as part of issue https://github.com/rancher-sandbox/rancher-desktop/issues/3862

File: pkg/rancher-desktop/backend/lima.ts [hunk 16]
```diff
@@ -1875,7 +1878,7 @@ CREDFWD_URL='http://${ hostIPAddr }:${ stateInfo.port }'
       throw ex;
     }
 
-    this.cfg = undefined;
+    this.cfg = defaultSettings;
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/lima.ts:1879]: This seems like a bad idea: we can end up running with the wrong configuration. This was previously optional to ensure that we fall over if we ended up somehow trying to start the backend without setting the configuration.

File: pkg/rancher-desktop/backend/mock.ts (hunk 6 above)
Comment (mook-as) [reply pkg/rancher-desktop/backend/mock.ts]: But… you made it a `RecursiveReadonly` instead of a `RecursivePartial`?

File: pkg/rancher-desktop/backend/wsl.ts [hunk 17]
```diff
@@ -1410,7 +1413,7 @@ export default class WSLBackend extends events.EventEmitter implements VMBackend
       if (await this.isDistroRegistered({ distribution: DATA_INSTANCE_NAME })) {
         await this.execWSL('--unregister', DATA_INSTANCE_NAME);
       }
-      this.cfg = undefined;
+      this.cfg = defaultSettings;
```
Comment (mook-as) [inline pkg/rancher-desktop/backend/wsl.ts:1414]: See line 137 above — this should be `undefined` here. (It should also be a separate commit.)

**Review by mook-as — CHANGES_REQUESTED** [review]
This PR is so large that I can only review commit-by-commit instead of the whole thing at once.  This review is only for the first commit (c4b1f0146e3f7f1a78b7c84285256e8da38b42eb).

Please move *everything* other than changing the structure of the settings and the settings version number to different commits, such that there is only mechanical changes other than those two.

File: pkg/rancher-desktop/config/settings.ts (hunk 10 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/config/settings.ts]: No, that's not it. But when I tried to export it from this module, the import failed unless I left the 'as const' part off. I don't know typescript well enough to understand what the problem is there.

**Comment by ericpromislow** [comment]
It looks like doing the work on a separate dev branch and then waiting to do a full review of merging all the work into `main` may not have been the best idea.

It'll take some time to reformulate this PR as a set of PRs again, but maybe that's what it will take. Here are the PRs I envision:

* move to v5 settings
* migrate from v4 to v5 settings
* drop the kim builder
* migrate the api from v0 to v1

**Comment by ericpromislow** [comment]
Dropping this PR and splitting this into three separate PRs, each of which will just target `main`...

File: background.ts [hunk 18]
```diff
@@ -973,7 +967,7 @@ class BackgroundCommandWorker implements CommandWorkerInterface {
   ): Promise<[string, string]> {
     const [needToUpdate, errors] = this.settingsValidator.validateTransientSettings(TransientSettings.value, newTransientSettings);
 
-    return Promise.resolve((() => {
+    return Promise.resolve((():[string, string] => {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply background.ts:983]: It's in a separate PR, but my IDE has been complaining about a type-mismatch the whole time. 

I know that the company behind this IDE is lagging behind the tslint/eslint we use, but it would be
nice to at least not have this warning.
